### ISSUE TITLE
Enhance `textDocument/references` to Utilize Multi-Core CPUs for Better Performance

### ISSUE THREAD

**Issue by eval-exec** [rust-lang/rust-analyzer#17753]
When invoking the `textDocument/references` action, the current implementation seems processes references sequentially on a single core. This approach is not utilizing the full potential of modern multi-core processors and leads to increased time cost for finding references in large codebases.

I propose enhancing the `textDocument/references` feature to utilize multi-core CPUs by parallelizing the reference search. Leveraging parallel computing (e.g., using crates like `rayon`) could significantly reduce the time needed for this operation.

I haven't deeply investigated the rust-analyzer codebase yet; is it possible for this feature to be implemented?

Thank you!

**Comment by davidbarsky** [comment]
yes, there is work being on this: https://github.com/rust-lang/rust-analyzer/issues/17491
